### PR TITLE
Cached sin and cos matrices for rotary at GPT-J model initialization for faster generation

### PR DESCRIPTION
**Pull request by kurumuz**
# What does this PR do?

Makes generations faster by caching sincos matrices for rotary for the max sequence length at model initialisation so it won't be done every other forward.

around ~%15 percent speedup overall
Tested on A100-SXM4-80GB
Benchmarks:
1 token forward average runtime out of 100 iterations with cached sin cos:
0.023421470640283642s

1 token forward average runtime out of 100 iterations without cached sin cos:
0.02738137301433869s

10 generations with 1 token context and 40 tokens generated without sincos caching:
1.410405054409057s

10 generations with 1 token context and 40 tokens generated with sincos caching:
1.2199230638332665s

Test script:
```
torch.manual_seed(123)
input_ids = torch.randint(0, 100, (1, 1)).cuda().long()

iterations = 11
with torch.no_grad():
    for i in range(iterations):
        if i == 1:
            #start counting time after tier 1 due to pytorch warming up
            t = time.perf_counter()
        outputs = model.generate(input_ids, max_length=input_ids.shape[-1] + 50, min_length=input_ids.shape[-1] + 50, do_sample=True, use_cache=True)
        #outputs = model.forward(input_ids)
print((time.perf_counter() - t) / (iterations - 1))
```

Models:
- GPT-J

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! Adding this buffer adds it to the state dict of the model, which then causes a lot of problems with loading/reloading. You might want to use the `persistent=False` option to have them not be saved.

**Comment by sgugger** [comment]
Mmm, now it looks like the tests are not running for some reason. Could you try refreshing your permissions as shown [here](https://support.circleci.com/hc/en-us/articles/360048210711-How-to-Refresh-User-Permissions-)? And/or push an empty commit?

File: src/transformers/models/gptj/modeling_gptj.py [hunk 1]
```diff
@@ -124,6 +124,10 @@ def __init__(self, config):
         if config.rotary_dim is not None:
             self.rotary_dim = config.rotary_dim
 
+        sin, cos = fixed_pos_embedding(x=torch.zeros(1, 1, self.rotary_dim), seq_len=max_positions)
+        self.register_buffer("sin", sin, persistent=False)
+        self.register_buffer("cos", cos, persistent=False)
```
Comment (ydshieh) [inline src/transformers/models/gptj/modeling_gptj.py:129]: Since you pass `seq_len=max_positions` to `fixed_pos_embedding`, the registered `sin` and `cos` should have seq. dim equal to `max_positions`. This changes the shapes of `sin` and `cos` before and after this PR.

This doesn't cause any issue?

File: src/transformers/models/gptj/modeling_gptj.py [hunk 2]
```diff
@@ -124,6 +124,10 @@ def __init__(self, config):
         if config.rotary_dim is not None:
             self.rotary_dim = config.rotary_dim
 
+        sin, cos = fixed_pos_embedding(x=torch.zeros(1, 1, self.rotary_dim), seq_len=max_positions)
```
Comment (ydshieh) [inline src/transformers/models/gptj/modeling_gptj.py:127]: If everything else works, before merge, it would be a good idea to explain why we can set the second `1` in `(1, 1, self.rotary_dim)` in a comment, so we don't need to read the `fixed_pos_embedding` just to figure this out.

Comment (ydshieh) [reply src/transformers/models/gptj/modeling_gptj.py:127]: `self.rotary_dim` could be `None`, and we will have to deal with this case too, otherwise the code will fail in such case.

File: src/transformers/models/gptj/modeling_gptj.py [hunk 3]
```diff
@@ -229,15 +233,15 @@ def forward(
...
+            
+            sincos = (self.sin, self.cos)
             k_rot = apply_rotary_pos_emb(k_rot, sincos, offset=offset)
             q_rot = apply_rotary_pos_emb(q_rot, sincos, offset=offset)
 
             key = torch.cat([k_rot, k_pass], dim=-1)
             query = torch.cat([q_rot, q_pass], dim=-1)
         else:
-            sincos = fixed_pos_embedding(key, 1, seq_len=seq_len)
+            sincos = (self.sin, self.cos)
```
Comment (ydshieh) [inline src/transformers/models/gptj/modeling_gptj.py:244]: This is what I mentioned above. This block is for `if self.rotary_dim is None`, and currently, this PR hasn't address this case in `__init__`
